pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBase.java

```diff
@@ -27,29 +27,32 @@
 import org.apache.pulsar.client.api.MessageId;
 import org.apache.pulsar.client.api.Producer;
 import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.Schema;
 import org.apache.pulsar.client.impl.conf.ProducerConfigurationData;
 
-public abstract class ProducerBase extends HandlerBase implements Producer {
+public abstract class ProducerBase<T> extends HandlerBase implements Producer<T> {
 
-    protected final CompletableFuture<Producer> producerCreatedFuture;
+    protected final CompletableFuture<Producer<T>> producerCreatedFuture;
     protected final ProducerConfigurationData conf;
+    protected final Schema<T> schema;
 
     protected ProducerBase(PulsarClientImpl client, String topic, ProducerConfigurationData conf,
-            CompletableFuture<Producer> producerCreatedFuture) {
+            CompletableFuture<Producer<T>> producerCreatedFuture, Schema<T> schema) {
         super(client, topic, new Backoff(100, TimeUnit.MILLISECONDS, 60, TimeUnit.SECONDS,
                 Math.max(100, conf.getSendTimeoutMs() - 100), TimeUnit.MILLISECONDS));
         this.producerCreatedFuture = producerCreatedFuture;
         this.conf = conf;
+        this.schema = schema;
     }
 
     @Override
-    public MessageId send(byte[] message) throws PulsarClientException {
-        return send(MessageBuilder.create().setContent(message).build());
+    public MessageId send(T message) throws PulsarClientException {
+        return send(MessageBuilder.create().setContent(schema.encode(message)).build());
     }
 
     @Override
-    public CompletableFuture<MessageId> sendAsync(byte[] message) {
-        return sendAsync(MessageBuilder.create().setContent(message).build());
+    public CompletableFuture<MessageId> sendAsync(T message) {
+        return sendAsync(MessageBuilder.create().setContent(schema.encode(message)).build());
```

Comment:
the below method `MessageId send(Message message)` should have the `Message` be generic as well. 

One problem that comes from that is that `MessageBuilder` can take a `ByteBuffer` as well as a `byte[]`, so the actual content can be off heap.

Reply:
That is only relevant when the there is no schema though.